### DISCUSSION TITLE
Create triggers with alembic and SQLAlchemy models

### DISCUSSION THREAD

**Discussion by RobinFrcd** [sqlalchemy/sqlalchemy#6806] [Usage Questions] [answered]
Hello,
I've a pretty complex project with a lot of SQLAlchemy models files I use to build alembic scripts.

I'd like to create some PSQL triggers. To keep things as clear as possible, I don't want to create the triggers in `env.py` or in `versions` files. I'd like to declare a trigger (or custom DDL) in my SQLAlchemy models files directly. Would it be possible to do this ?

Thanks

**Comment by zzzeek** [comment, marked as answer]
the CREATE TRIGGER calls will be in the version files regardless as long as you are using Alembic.  there's an extension for Alembic that supports PG triggers and SPs at https://olirice.github.io/alembic_utils/.

**Reply by RobinFrcd** [reply]
Thank you for your answer! So there's no way to declare the triggers in the models files, that's right ? Looks like I'll need to do a `register_entities` in the `env.py` file.

**Reply by zzzeek** [reply]
you can declare triggers with your models using the [DDL()](https://docs.sqlalchemy.org/en/14/core/ddl.html#controlling-ddl-sequences) construct:

```py
class MyModel(Base):
    __table__ = 'mytable'

    id = Column('id', Integer, primary_key=True)
    data = Column('data', String(50))


trigger = DDL(
    "CREATE TRIGGER dt_ins BEFORE INSERT ON mytable "
    "FOR EACH ROW EXECUTE PROCEDURE my_func();"
)


event.listen(
    MyModel.__table__,
    'after_create',
    trigger.execute_if(dialect='postgresql')
)
```

that won't have anything to do with Alembic though, only if you were to call metadata.create_all().

it was my impression that the alembic-utils project combines the above metadata-based approach with an autogenerate scheme that is inspired by the [replaceable object recipe](https://alembic.sqlalchemy.org/en/latest/cookbook.html#replaceable-objects), but i haven't really looked at it.

**Reply by RobinFrcd** [reply]
Thank you for your answer, that's what I tried first, but sadly it doesn't add the triggers to the migration scripts when I run alembic. :/

**Reply by zzzeek** [reply]
correct, that's autogenerate support, for which you would need to use the above mentioned alembic-utils to achieve.
